Mathlib/AlgebraicGeometry/AffineSpace.lean

```diff
@@ -0,0 +1,362 @@
+
+@[simp]
+lemma isoOfIsAffine_hom_app_top [IsAffine S] :
+    (isoOfIsAffine n S).hom.app ⊤ =
+      (Scheme.ΓSpecIso _).hom ≫ eval₂Hom ((𝔸(n; S) ↘ S).app ⊤) (coord S) := by
+  simp [isoOfIsAffine_hom]
+
+@[local simp]
+lemma isoOfIsAffine_inv_app_top_coord [IsAffine S] (i) :
+    (isoOfIsAffine n S).inv.app ⊤ (coord _ i) = (Scheme.ΓSpecIso (.of _)).inv (.X i) :=
+  homOfVector_app_top_coord _ _ _
+
+@[reassoc (attr := simp)]
+lemma isoOfIsAffine_inv_over [IsAffine S] :
+    (isoOfIsAffine n S).inv ≫ 𝔸(n; S) ↘ S = Spec.map C ≫ S.isoSpec.inv :=
+  pullback.lift_fst _ _ _
+
+instance [IsAffine S] : IsAffine 𝔸(n; S) := isAffine_of_isIso (isoOfIsAffine n S).hom
+
+variable (n) in
+/-- The affine space over an affine base is isomorphic to the spectrum of the polynomial ring. -/
+def SpecIso (R : CommRingCat.{max u v}) :
+    𝔸(n; Spec R) ≅ Spec (.of (MvPolynomial n R)) :=
+  isoOfIsAffine _ _ ≪≫ Scheme.Spec.mapIso (MvPolynomial.mapEquiv _
+    (Scheme.ΓSpecIso R).symm.commRingCatIsoToRingEquiv).toCommRingCatIso.op
+
+@[simp]
+lemma SpecIso_hom_app_top (R : CommRingCat.{max u v}) :
+    (SpecIso n R).hom.app ⊤ = (Scheme.ΓSpecIso _).hom ≫
+      eval₂Hom ((Scheme.ΓSpecIso _).inv ≫ (𝔸(n; Spec R) ↘ Spec R).app ⊤) (coord (Spec R)) := by
+  simp only [SpecIso, Iso.trans_hom, Functor.mapIso_hom, Iso.op_hom, RingEquiv.toCommRingCatIso_hom,
+    RingEquiv.toRingHom_eq_coe, Scheme.Spec_map, Quiver.Hom.unop_op, Scheme.comp_coeBase,
+    TopologicalSpace.Opens.map_comp_obj, TopologicalSpace.Opens.map_top, Scheme.comp_app,
+    isoOfIsAffine_hom_app_top]
+  erw [Scheme.ΓSpecIso_naturality_assoc]
+  congr 1
+  apply ringHom_ext'
+  · ext; simp; rfl
+  · simp
+
+@[local simp]
+lemma SpecIso_inv_app_top_coord (R : CommRingCat.{max u v}) (i) :
+    (SpecIso n R).inv.app ⊤ (coord _ i) = (Scheme.ΓSpecIso (.of _)).inv (.X i) := by
+  simp only [SpecIso, Iso.trans_inv, Functor.mapIso_inv, Iso.op_inv, RingEquiv.toCommRingCatIso_inv,
+    mapEquiv_symm, RingEquiv.toRingHom_eq_coe, Scheme.Spec_map, Quiver.Hom.unop_op,
+    Scheme.comp_coeBase, TopologicalSpace.Opens.map_comp_obj, TopologicalSpace.Opens.map_top,
+    Scheme.comp_app, CommRingCat.comp_apply]
+  erw [isoOfIsAffine_inv_app_top_coord, ← CommRingCat.comp_apply]
+  rw [← Scheme.ΓSpecIso_inv_naturality]
+  erw [CommRingCat.comp_apply]
+  congr 1
+  exact map_X _ _
+
+@[reassoc (attr := simp)]
+lemma SpecIso_inv_over (R : CommRingCat.{max u v}) :
+    (SpecIso n R).inv ≫ 𝔸(n; Spec R) ↘ Spec R = Spec.map C := by
+  simp only [SpecIso, Iso.trans_inv, Functor.mapIso_inv, Iso.op_inv, RingEquiv.toCommRingCatIso_inv,
+    mapEquiv_symm, RingEquiv.toRingHom_eq_coe, Scheme.Spec_map, Quiver.Hom.unop_op, Category.assoc,
+    isoOfIsAffine_inv_over, Scheme.isoSpec_Spec_inv, ← Spec.map_comp]
+  congr 1
+  rw [Iso.inv_comp_eq]
+  ext
+  exact map_C _ _
+
+section functorial
+
+variable (n) in
+/-- `𝔸(n; S)` is functorial wrt `S`. -/
+def map {S T : Scheme.{max u v}} (f : S ⟶ T) : 𝔸(n; S) ⟶ 𝔸(n; T) :=
```

Comment:
Do you think it is reasonable to have it as a functor?

Reply:
Added.